File: cowichan/baSrcPaper/go/chain/par/src/all/product.go
/*
 * product: matrix-vector product
 *
 * input:
 *   nelts: the number of elements
 *   Outer_matrix: the real matrix
 *   Outer_vector: the real vector
 *
 * output:
 *   Product_result: a real vector, whose values are the result of the product
 */
package all

var Product_result []Double;

func fill_result_impl(begin, end, ncols int, done chan bool) {
  if (begin + 1 == end) {
    var sum Double = 0;
    for j := 0; j < ncols; j++ {
      sum += Outer_matrix[begin][j] * Outer_vector[j];
    }
    Product_result[begin] = sum;
    done <- true
  } else {
    middle := begin + (end - begin) / 2;
    go fill_result_impl(begin, middle, ncols, done);
    fill_result_impl(middle, end, ncols, done);
  }
}

func fill_result(nrows, ncols int) {
  done := make(chan bool);
  // parallel for on rows
  go fill_result_impl(0, nrows, ncols, done);
  for i := 0; i < nrows; i++ {
    <-done;
  }
}

func Product(nelts int) {
  Product_result = make ([]Double, nelts)
  fill_result(nelts, nelts);
}

File: cowichan/baSrcPaper/cilk/chain/par/product.cpp
/*
 * product: matrix-vector product
 *
 * input:
 *   nelts: the number of elements
 *   outer_matrix: a real matrix
 *   outer_vector: a real vector
 *
 * output:
 *   product_result: a real vector, whose values are the result of the product
 */

#include <cilk/cilk.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

extern double *outer_matrix;
extern double *outer_vector;
double *product_result;

// parallel for on [begin, end)
void fill_result(int begin, int end, int ncols) {
  int middle = begin + (end - begin) / 2;
  double sum = 0;
  int j;
  if (begin + 1 == end) {
    for (j = 0; j < ncols; j++) {
      sum += outer_matrix[begin*ncols + j] * outer_vector[j];
    }
    product_result[begin] = sum;
    return;
  }
  cilk_spawn fill_result(begin, middle, ncols);
  cilk_spawn fill_result(middle, end, ncols);
}

void product(int nelts) {
  product_result = (double*) malloc (sizeof (double) * nelts);
  cilk_spawn fill_result(0, nelts, nelts);
}
